helix-core/src/main/java/org/apache/helix/controller/rebalancer/topology/Topology.java

```diff
@@ -81,62 +75,43 @@ public Topology(final List<String> allNodes, final List<String> liveNodes,
       throw new HelixException(String.format("Config for instances %s is not found!",
           _allInstances.removeAll(_instanceConfigMap.keySet())));
     }
-
     _clusterConfig = clusterConfig;
-    _types = new LinkedHashSet<>();
-    _topologyAwareEnabled = clusterConfig.isTopologyAwareEnabled();
 
-    if (_topologyAwareEnabled) {
+    if (clusterConfig.isTopologyAwareEnabled()) {
       String topologyDef = _clusterConfig.getTopology();
       if (topologyDef != null) {
         // Customized cluster topology definition is configured.
-        String[] types = topologyDef.trim().split("/");
-        for (int i = 0; i < types.length; i++) {
-          if (types[i].length() != 0) {
-            _types.add(types[i]);
+        String[] topologyStr = topologyDef.trim().split("/");
+        int lastValidTypeIdx = topologyStr.length - 1;
+        while(lastValidTypeIdx >= 0) {
+          if (topologyStr[lastValidTypeIdx].length() != 0) {
+            break;
           }
+          --lastValidTypeIdx;
         }
-        if (_types.size() == 0) {
-          logger.error("Invalid cluster topology definition " + topologyDef);
+        if (lastValidTypeIdx < 0) {
           throw new HelixException("Invalid cluster topology definition " + topologyDef);
-        } else {
-          String lastType = null;
-          for (String type : _types) {
-            _defaultDomainPathValues.put(type, "Helix_default_" + type);
-            lastType = type;
-          }
-          _endNodeType = lastType;
-          _faultZoneType = clusterConfig.getFaultZoneType();
-          if (_faultZoneType == null) {
-            _faultZoneType = _endNodeType;
-          }
-          if (!_types.contains(_faultZoneType)) {
-            throw new HelixException(String
-                .format("Invalid fault zone type %s, not present in topology definition %s.",
-                    _faultZoneType, topologyDef));
-          }
-          _useDefaultTopologyDef = false;
+        }
+        _endNodeType = topologyStr[lastValidTypeIdx];
+        _faultZoneType = clusterConfig.getFaultZoneType();
+        if (_faultZoneType == null) {
+          _faultZoneType = _endNodeType;
+        }
+        if (Arrays.stream(topologyStr).noneMatch(type -> type.equals(_faultZoneType))) {
```

Comment:
Please correct me if I am wrong. 
Comparing to previous approach (set. contains()), the stream.noneMatch() does introduced a o(n)worst case complexity. 
However, since there is no new LinkedHashSet created, it saved a O(n) time for constructing the LinkedHashSet and a O(n) space complexity for the object itself. 

It is better to pass the already parsed array to computeInstanceTopologyMap so we do not compute again in line 274. However, I think we could just use the original array topologyStr, no need to introduce another LinkedHashSet.

Reply:
One more optimization. We could put this check in the else branch for line 97. If we already set the _faultZoneType to be the same as _endNodeType, then _faultZoneType must belongs to topologyStr.

Reply:
Considering the case "/domain/rack///node/", you will need to process the array. So either you have a new array or a LinkedHashSet. The latter can save us some time here.
I was thinking about these 2 points together. And my conclusion is that keep the original logic is better : )

Reply:
Had an offline discussion. Although the time consumed here in the constructor is the same, it is better to have a smaller loop with no empty string when creating the domain k-v map for all instance. Having extra empty string in the for loop may introduce performance issue considering the number of instance.